extensions/arc/deployment/src/main/java/io/quarkus/arc/deployment/SyntheticBeanBuildItem.java

```diff
@@ -79,6 +84,8 @@ public static class ExtendedBeanConfigurator extends BeanConfiguratorBase<Extend
         private Function<SyntheticCreationalContext<?>, ?> fun;
         private boolean staticInit;
 
+        private Runnable checkActive;
```

Comment:
It would be possible to pass `SyntheticCreationalContext`, I think. Are the dependencies registered as injection points (and therefore available from `SyntheticCreationalContext.getInjectedReference()`)?

Other than that, `checkActive` is called during `Bean.create()`, which means ArC should already have been booted up, so `Arc.cotnainer()` should be safe to use. But I'm not 100% sure, so don't quote me on that :-)

Reply:
> It would be possible to pass SyntheticCreationalContext, I think. Are the dependencies registered as injection points (and therefore available from SyntheticCreationalContext.getInjectedReference())?

In my case they are, yes. It seems like a reasonable requirement as far as I'm concerned.

Reply:
OK, I'll try to add this.

Reply:
I'm looking into this and I'm wondering, when you say

> The check might rely on other (startup-initialized) beans

does that mean you need to call methods on the other beans, or you need to check that the other beans are also active? Or both?

If you need to check that the other beans are also active, then using a `SyntheticCreationalContext` won't work, at least not always. The SCC contains contextual references for the injected beans, which means contextual instances for pseudo-scoped beans. If any of those is not active, we'll throw `InactiveBeanException` when creating the SCC. If the bean is normal-scoped, the contextual reference is a client proxy, so we'll only throw `InactiveBeanException` when calling a method, which will be in the active check procedure, so bearable, but probably not very useful still.

Reply:
> does that mean you need to call methods on the other beans, or you need to check that the other beans are also active? Or both?

In my case I just need to check that other beans are active. I could see calling methods on other beans being useful, but I don't have a use case for that right now.

> If you need to check that the other beans are also active, then using a SyntheticCreationalContext won't work, at least not always. The SCC contains contextual references for the injected beans, which means contextual instances for pseudo-scoped beans.

The beans I'm interested in are application-scoped.

> If any of those is not active, we'll throw InactiveBeanException when creating the SCC.

Uh, I didn't anticipate that. Aren't synthetic consumers supposed to be able to work even if their dependencies are inactive?

> If the bean is normal-scoped, the contextual reference is a client proxy, so we'll only throw InactiveBeanException when calling a method, which will be in the active check procedure, so bearable, but probably not very useful still.

I suppose I can use `ClientProxy.unwrap()`? But yeah I agree, it's a bit clunky.

Anyway... If I understand correctly, it would be a better idea to use `Arc.container()` to retrieve beans in `checkActive` and then [any of the ways we discussed for checking if beans are active](https://github.com/quarkusio/quarkus/pull/41810#discussion_r1679632971) instead?... right? In which case maybe exposing `SyntheticCreationalContext` there is not a good idea.

To be honest I was suggesting to expose `SyntheticCreationalContext` because I thought this was a "better" way to retrieve beans, but I have no idea why it's better. So maybe it isn't, at least in this case.

Reply:
> Aren't synthetic consumers supposed to be able to work even if their dependencies are inactive?

I guess that really depends on what the consumer does. The SCC contains [eagerly initialized] contextual references, so that cannot work in case a pseudo-scoped bean might be inactive. It would work for normal-scoped beans, but as we agreed, it's a bit weird. See below for a possibly better idea.

> I suppose I can use `ClientProxy.unwrap()`? But yeah I agree, it's a bit clunky.

Yeah, that's correct, `ClientProxy.unwrap()` would work.

> To be honest I was suggesting to expose `SyntheticCreationalContext` because I thought this was a "better" way to retrieve beans, but I have no idea why it's better. So maybe it isn't, at least in this case.

I guess that's true in this case. SCC gives you a contextual reference to a bean, while you'd need a `Bean` or `InjectableInstance` or something like that. So `Arc.container()` seems better, at least until we figure out a better replacement.

Maybe we could do something like

```java
interface ActiveCheckContext {
    <T> InstanceHandle<T> getInjectedBean(Class<T> requiredType, Annotation... qualifiers);
    <T> InstanceHandle<T> getInjectedBean(TypeLiteral<T> requiredType, Annotation... qualifiers);
}
```

We would reuse the injection points configured via `BeanConfigurator.addInjectionPoint()`, but instead of SCC that gives you contextual references, we'd give you lazily initialized `InstanceHandle`. From there, you could obtain the `InjectableBean` to call `isActive()`, or perhaps obtain the contextual reference to call methods.

Reply:
FWIW using `Arc.container()` to retrieve beans works perfectly fine in my WIP PR; marking this thread as resolved.